ykrt/src/trace/sw/mod.rs

```diff
@@ -0,0 +1,120 @@
+//! Software tracer.
+
+use crate::frame::BitcodeSection;
+
+use super::{errors::InvalidTraceError, TraceCollector, TraceIterator, TracedAOTBlock};
+use std::{cell::RefCell, collections::HashMap, error::Error, ffi::CString, sync::Arc};
+
+#[derive(Debug, Eq, PartialEq)]
+struct BB {
+    function_index: u32,
+    block_index: u32,
+}
+
+thread_local! {
+    static BASIC_BLOCKS: RefCell<Vec<BB>> = RefCell::new(vec![]);
+}
+
+extern "C" {
+    fn get_function_names(
+        section: *const BitcodeSection,
+        vec: *const libc::c_uint,
+        vec_size: usize,
+        result: *mut *mut IRFunctionNameIndex,
+        len: *mut libc::size_t,
+    );
+    fn free_function_names(funcNames: *mut IRFunctionNameIndex);
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct IRFunctionNameIndex {
+    pub index: u32, //libc::c_int,
+    pub name: *const libc::c_char,
+}
+/// Inserts LLVM IR basicblock metadata into a thread-local BASIC_BLOCKS vector.
+///
+/// # Arguments
+/// * `function_index` - The index of the function to which the basic block belongs.
+/// * `block_index` - The index of the basic block within the function.
+pub fn trace_basicblock(function_index: u32, block_index: u32) {
+    BASIC_BLOCKS.with(|v| {
+        v.borrow_mut().push(BB {
+            function_index,
+            block_index,
+        });
+    })
+}
+
+pub(crate) struct SWTracer {}
+
+impl super::Tracer for SWTracer {
+    fn start_collector(self: Arc<Self>) -> Result<Box<dyn TraceCollector>, Box<dyn Error>> {
+        return Ok(Box::new(SWTTraceCollector {}));
+    }
+}
+
+impl SWTracer {
+    pub fn new() -> Result<Self, Box<dyn Error>> {
+        Ok(SWTracer {})
+    }
+}
+
+struct SWTTraceCollector {}
+
+impl TraceCollector for SWTTraceCollector {
+    fn stop_collector(self: Box<Self>) -> Result<Box<dyn TraceIterator>, InvalidTraceError> {
+        let mut aot_blocks: Vec<TracedAOTBlock> = vec![];
+        BASIC_BLOCKS.with(|mtt| {
+            let mut func_store: HashMap<u32, CString> = HashMap::new();
+            let func_indices: Vec<u32> = mtt.borrow().iter().map(|x| x.function_index).collect();
+            let mut func_names: *mut IRFunctionNameIndex = std::ptr::null_mut();
+            let mut func_names_len: libc::size_t = 0;
+            let bc_section = crate::compile::jitc_llvm::llvmbc_section();
+            unsafe {
+                get_function_names(
```

Comment:
@vext01 here's my first attemp on obtaining funciton names by inxes using `ykllvmwrap` module.
When runned as:
```shell
YKB_TRACER=sw cargo test  ::simple_swt.c -- --nocapture 
```
It constructs the following trace:
```text
Mapped { func_name: "main", bb: 10 }
Mapped { func_name: "inc", bb: 0 }
Mapped { func_name: "inc", bb: 1 }
Mapped { func_name: "add1", bb: 0 }
Mapped { func_name: "add1", bb: 1 }
Mapped { func_name: "add1", bb: 3 }
Mapped { func_name: "add1", bb: 2 }
Mapped { func_name: "inc", bb: 2 }
Mapped { func_name: "inc", bb: 4 }
Mapped { func_name: "inc", bb: 3 }
Mapped { func_name: "main", bb: 11 }
Mapped { func_name: "main", bb: 8 }
Mapped { func_name: "main", bb: 9 }
```